server/src/main/java/org/elasticsearch/monitor/jvm/JvmGcMonitorService.java

```diff
@@ -165,13 +165,11 @@ public JvmGcMonitorService(Settings settings, ThreadPool threadPool) {
     }
 
     private static TimeValue getValidThreshold(Settings settings, String key, String level) {
-        TimeValue threshold = settings.getAsTime(level, null);
-        if (threshold == null) {
+        final String thresholdString = settings.get(level, null);
+        if (thresholdString == null) {
             throw new IllegalArgumentException("missing gc_threshold for [" + getThresholdName(key, level) + "]");
         }
-        if (threshold.nanos() <= 0) {
-            throw new IllegalArgumentException("invalid gc_threshold [" + threshold + "] for [" + getThresholdName(key, level) + "]");
-        }
+        TimeValue threshold = TimeValue.parseTimeValue(thresholdString, null, getThresholdName(key, level));
```

Comment:
I'm not sure that I like that we are manually parsing settings here. I understand that we're doing it to have the full setting name. How about:

```diff
diff --git a/server/src/main/java/org/elasticsearch/monitor/jvm/JvmGcMonitorService.java b/server/src/main/java/org/elasticsearch/monitor/jvm/JvmGcMonitorService.java
index 5744814293f..cad01bd9856 100644
--- a/server/src/main/java/org/elasticsearch/monitor/jvm/JvmGcMonitorService.java
+++ b/server/src/main/java/org/elasticsearch/monitor/jvm/JvmGcMonitorService.java
@@ -165,11 +165,15 @@ public class JvmGcMonitorService extends AbstractLifecycleComponent {
     }
 
     private static TimeValue getValidThreshold(Settings settings, String key, String level) {
-        final String thresholdString = settings.get(level, null);
-        if (thresholdString == null) {
+        final TimeValue threshold;
+        try {
+            threshold = settings.getAsTime(level, null);
+        } catch (final IllegalArgumentException e) {
+            throw new IllegalArgumentException("failed to parse [" + getThresholdName(key, level) + "]", e);
+        }
+        if (threshold == null) {
             throw new IllegalArgumentException("missing gc_threshold for [" + getThresholdName(key, level) + "]");
         }
-        TimeValue threshold = TimeValue.parseTimeValue(thresholdString, null, getThresholdName(key, level));
         return threshold;
     }
```

and the necessary update to the test (assert the message, assert the cause, assert the message on the cause)?

Reply:
I've replaced this with something very similar to what you suggested, catching `RuntimeException` as not much else could go wrong trying to get the setting other than a parsing failure.